drivers/platform/xilinx/i2c.c

```diff
@@ -57,65 +67,98 @@
  * @return SUCCESS in case of success, FAILURE otherwise.
  */
 int32_t i2c_init(struct i2c_desc **desc,
-		 const struct i2c_init_param *param)
+		 const struct i2c_init_param *init_param)
 {
-#ifdef XIIC_H
-	i2c_desc *dev;
-	xil_i2c_desc *xil_dev;
-	xil_i2c_init_param *xil_param;
-	int32_t ret;
-
-	dev = calloc(1, sizeof *dev);
-	if(!dev)
-		return FAILURE;
+	int32_t		ret;
+	i2c_desc	*idesc;
+	xil_i2c_desc	*xdesc;
+	xil_i2c_init	*xinit;
 
-	dev->extra = calloc(1, sizeof *xil_dev);
-	if(!(dev->extra)) {
-		free(dev);
+	idesc = (struct i2c_desc *)malloc(sizeof(*idesc));
+	xdesc = (struct xil_i2c_desc *)malloc(sizeof(*xdesc));
+	if(!idesc || !xdesc) {
+		free(idesc);
+		free(xdesc);
 		return FAILURE;
 	}
 
-	xil_dev = dev->extra;
-	xil_param = param->extra;
+	idesc->max_speed_hz = init_param->max_speed_hz;
+	idesc->slave_address = init_param->slave_address;
+	xinit = init_param->extra;
+	idesc->extra = xdesc;
 
-	xil_dev->type = xil_param->type;
-	xil_dev->id = xil_param->id;
-	dev->max_speed_hz = param->max_speed_hz;
-	dev->slave_address = param->slave_address;
+	switch (xinit->type) {
+	case IIC_PL:
+#ifdef XIIC_H
+		xdesc->instance = (XIic *)malloc(sizeof(*(xdesc->instance)));
+		if(!xdesc->instance)
+			goto error;
 
-	xil_dev->config = XIic_LookupConfig(xil_dev->id);
-	if (xil_dev->config == NULL)
-		goto error;
+		xdesc->config = XIic_LookupConfig(xinit->device_id);
+		if(xdesc->config == NULL)
+			goto error;
 
-	ret = XIic_CfgInitialize(&xil_dev->instance, xil_dev->config,
-				 xil_dev->config->BaseAddress);
-	if(ret != 0)
-		goto error;
+		ret = XIic_CfgInitialize(xdesc->instance,
+					 xdesc->config,
+					 ((XIic_Config*)xdesc->config)
+					 ->BaseAddress);
+		if(ret != SUCCESS)
+			goto error;
 
-	Xil_ExceptionInit();
+		Xil_ExceptionInit();
+		Xil_ExceptionRegisterHandler(XIL_EXCEPTION_ID_IRQ_INT,
+					     XIic_InterruptHandler,
+					     &xdesc->instance);
```

Comment:
This part is a bit tricky. If the core should be the only one to generate interrupts in the system it is OK to connect it directly to the ALU interrupt handler. But if there are more cores need to generate interrupts there would need to be an interrupt controller that will be connected to the ALU handler and all the other ones are connected to the interrupt controller. For example when using the I2C core and a timer, or an UART core.

Maybe the driver could check if there is an interrupt controller available, and if there is connect to it instead, and if there is not then do what it does now. Maybe this check can be done with and XPAR_NUM_OF_INSTANCES like the availability of the I2C cores?